accepted/0003-interactive-audit-resolver.md

```diff
@@ -0,0 +1,165 @@
+# Interactive audit resolver
+
+## Summary
+
+This proposal is for adding means for a human to resolve issues from `npm audit` and interactively make decisions about each issue. Available as `npm audit resolve` command.
+
+Features:
+- Remember the resolutions and allow them to affect `npm audit` behavior.
+- Decision options include 'ignore', 'remind later', 'fix', 'remove'.
+- Allow tracking who resolved an issue and when using git history.
+
+
+## Motivation
+
+At times, running `npm audit fix` won't fix all audit issues. Fixes might not be available or the user might not want to apply some of them, eg. major version updates or build/test dependencies.
+
+It should be possible to make `npm audit` a step in a CI pipeline to go red every time a new vulnerability affects the project. Managing security of dependencies should be quick to update, effective, easy to audit over time and secure in itself.
+
+For that, the user needs a new tool which makes addressing issues one by one convenient even across a whole ecosystem of projects.
+
+## Detailed Explanation
+
+`npm audit resolve` runs audit and iterates over all actions from the output with a prompt for a decision on each. 
+If fix is available, it's offered as one of the options to proceed. Other options include: ignore, remind later and delete.
+
+All decisions are stored in `audit-resolve.json` file as key-value, where key is `${advisory id}|${dependency path}` and value is:
```

Comment:
@isaacs I'm under the impression that the `${dependencyPath}` bit here is very non-arborist and not necessary overall in the context of an audit triage system, would like to hear your thoughts on it

Reply:
I'm not a native speaker and I'm not familiar enough with the culture to grasp what you mean exactly when calling something non-arborist. I'm going to focus on the "~necessary" part.

Mixing advisory and dependency path as a key in the audit-resolve file is the core of the whole idea.

It's the first time I was able to safely ignore vulnerabilities I knew won't affect me.

_A story:
My adventure with checking dependencies started with adding `nsp check` to our projects. 
The number of reports in software we used was low, so I didn't know what I'm getting myself into when I made `nsp check` a step in CI.
REDoS vulnerabilities got very popular after that and it's been an annoyance. nsp allowed creating an ignore list with just the advisoty id, so after finding out that a 5th level dependency of an old testing tool we use (or something) is unlikely to ever be fixed, we had to ignore all REDoS cases in all 20+ projects. 
Then `npm audit` showed up so I switched. It had no means of ignoring anything. CI step was removed and I started working on a wrapper that ended up being the npm-audit-resolver._

The key features I value in this solution:
- if I ignore the vulnerability in one package, it won't transfer to another
- if I ignore a vulnerability in a package coming in as a dependency of a dependency I don't really invoke in the app, I'm not automatically ignoring it elsewhere.
- if a new instance of the vulnerability shows up in dependencies of something else or in a dep tree of newly installed dependency - it doesn't get ignored 

While ignoring vulnerabilities seems like a thing that doesn't improve security when you hear about it, in this case it's a tool to really improve visibility of the reports you should actually look into and **increase** security. 

Ignoring vulnerability report is not here for convenience, but to make `npm audit` a viable step in the CI pipeline.

Reply:
ah 😄 sorry about that, I meant https://github.com/npm/arborist